k2/csrc/fsa_algo.cu

```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
   }
 }
 
+// TODO(fangjun): compute arc_map
+Fsa Union(FsaVec &fsas, Array1<int32_t> *arc_map /*= nullptr*/) {
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+
+  ContextPtr &context = fsas.Context();
+  const int32_t *fsas_row_splits1_data = fsas.RowSplits(1).Data();
+  const int32_t *fsas_row_splits2_data = fsas.RowSplits(2).Data();
+  const int32_t *fsas_row_ids1_data = fsas.RowIds(1).Data();
+  const int32_t *fsas_row_ids2_data = fsas.RowIds(2).Data();
+  const Arc *arcs_data = fsas.values.Data();
+
+  int32_t num_fsas = fsas.Dim0();
+  int32_t num_states = fsas.TotSize(1);
+  int32_t num_arcs = fsas.TotSize(2);
+
+  // A new start state and a new final state are added (+2).
+  // The final state of each fsa is removed (-num_fsas)
+  int32_t num_out_states = num_states + 2 - num_fsas;
```

Comment:
what happened here if one of input Fsa is empty?

Reply:
It will abort the program.

It indeed checks for the empty case
https://github.com/k2-fsa/k2/blob/c1e16fc0ea65b5f7be322291b6852d465b44637b/k2/csrc/fsa_algo.cu#L581


The empty case is not supported as the precondition is documented clearly
in the comment  of this function
https://github.com/k2-fsa/k2/blob/c1e16fc0ea65b5f7be322291b6852d465b44637b/k2/csrc/fsa_algo.h#L244-L245

Reply:
Oh, seems I miss your documentation here, sorry. But as @danpovey said below, I now also think we should not non=empty requirements on such basic algorithms, we should handle such cases in it.